intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java

```diff
@@ -0,0 +1,75 @@
+import com.intellij.openapi.project.Project;
+import com.intellij.util.messages.MessageBusConnection;
+import nl.tudelft.watchdog.core.logic.event.TrackingEventManager;
+import nl.tudelft.watchdog.core.logic.event.eventtypes.staticanalysis.StaticAnalysisType;
+import nl.tudelft.watchdog.core.logic.ui.listeners.CoreMarkupModelListener;
+import org.jetbrains.annotations.NotNull;
+
+public class IntelliJMarkupModelListener extends CoreMarkupModelListener implements MarkupModelListener, Disposable {
+
+    private IntelliJMarkupModelListener(TrackingEventManager trackingEventManager) {
+        super(trackingEventManager);
+    }
+
+    public static IntelliJMarkupModelListener initializeAfterAnalysisFinished(
+            Project project, Disposable disposable, Document document, TrackingEventManager trackingEventManager) {
+
+        final IntelliJMarkupModelListener markupModelListener = new IntelliJMarkupModelListener(trackingEventManager);
+
+        // We need to run this in smart mode, because the very first time you start your editor, it is very briefly
+        // in dumb mode and the codeAnalyzer thinks (incorrectly) it is  finished.
+        // Therefore, wait for smart mode and only then start listening, to make sure the codeAnalyzer actually did its thing.
+        DumbServiceImpl.getInstance(project).runWhenSmart(() -> {
+            final DaemonCodeAnalyzerImpl analyzer = (DaemonCodeAnalyzerImpl) DaemonCodeAnalyzer.getInstance(project);
+            final MessageBusConnection messageBusConnection = project.getMessageBus().connect(disposable);
+
+            messageBusConnection.subscribe(DaemonCodeAnalyzer.DAEMON_EVENT_TOPIC, new DaemonCodeAnalyzer.DaemonListenerAdapter() {
+                @Override
+                public void daemonFinished() {
+                    // We only receive global Code Analyzer events. This means that once such an event triggered,
+                    // we do not know for which file the analysis actually succeeded. Therefore we need to check
+                    // the so-called "dirty" state of the file for this particular analyzer to check whether
+                    // the analyzer actually finished. In this case, `null` indicates: the file is not dirty.
+                    if (analyzer.getFileStatusMap().getFileDirtyScope(document, Pass.UPDATE_ALL) == null) {
+                        final MarkupModelImpl markupModel = (MarkupModelImpl) DocumentMarkupModel.forDocument(document, project, true);
+                        markupModel.addMarkupModelListener(disposable, markupModelListener);
+                        messageBusConnection.disconnect();
+                    }
+                }
+            });
+        });
+
+        return markupModelListener;
+    }
+
+    @Override
+    public void afterAdded(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        if (rangeHighlighterEx.getLayer() == HighlighterLayer.WARNING) {
+            addCreatedWarning(StaticAnalysisType.UNKNOWN);
```

Comment:
Instead of flushing every single warning, we could use the Eclipse model of processing all warnings on save. It might be possible with http://arhipov.blogspot.nl/2011/04/code-snippet-intercepting-on-save.html to call the `add{Created,Removed}Warning()`. Then whenever warnings are removed, we have to clear up our internal storage of these two files and then "flush" when the save action occurs. WDYT @Inventitech ?

Reply:
We discussed this in a meeting